cpp/src/arrow/util/hashing.h

```diff
@@ -428,6 +428,22 @@ class ScalarMemoTable : public MemoTable {
         value, [](int32_t i) {}, [](int32_t i) {}, out_memo_index);
   }
 
+  Status MaybeInsert(const Scalar& value) {
```

Comment:
Looks `GetOrInsert` can do the job?

Reply:
ah, yeah I think it can. It just seemed like it might be nice to have a slightly simpler code path. but I can drop this and use GetOrInsert if we prefer

Reply:
Or at least implement this new API with `GetOrInsert` — the `Status GetOrInsert(const Scalar& value, int32_t* out_memo_index)` version should generate the same code more or less as this (with the exception of the unused index)

Reply:
I guess wrapping `GetOrInsert` would at least mean someone else doesn't have to make a dummy return value. I was also thinking there might be some benefits to not calling empty closures on each value. I'm not sure that's the case, but curious on your thoughts @wesm  @cyb70289

Reply:
Wrapping with a new API looks more clear.
I'm not worried about the empty closures and dummy return value. Probably compiler will optimize them to nothing if it decides to inline `GetOrInsert` as it's small.
A quick example, there's no unnecessary function calls or value assignments
https://godbolt.org/z/TjjTxzz8j

Reply:
Indeed, the empty closures will be easily optimized away by the compiler as dead / side-effect-free code

Reply:
thanks! I will make the change, and I appreciate the insight (and reference)